File: examples/transactions/v1/javascript/offlineTransactionAuthorization.js
const algosdk = require('/Users/ryanrfox/algorand/docs/testing/js/node_modules/algosdk') //TODO:'algosdk');
const fs = require('fs');
var client = null;
// make connection to node
async function setupClient() {
    if( client == null){
        const token = "a31f09a18dbf7ad68c9e0ff22355774fb89c67ed2c4642d6c6822f9360cd7697"//TODO:"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
        const server = "http://localhost";
        const port = 49392 //TODO:4001;
        let algodClient = new algosdk.Algod(token, server, port);
        client = algodClient;
    } else {
        return client;
    }
    return client;
}
// recover account for example
function recoverAccount(){
    const passphrase = "boy kidney fall hamster ecology mercy inquiry vast deal normal vibrant labor couch economy embody glory possible color burger addict soap almost margin about negative"; // TODO:"Your 25-word mnemonic goes here";"
    let myAccount = algosdk.mnemonicToSecretKey(passphrase);
    return myAccount;
}
// Function used to wait for a tx confirmation
var waitForConfirmation = async function(algodclient, txId) {
    while (true) {
        let lastround = (await algodclient.status()).lastRound;
        let pendingInfo = await algodclient.pendingTransactionInformation(txId);
        if (pendingInfo.round != null && pendingInfo.round > 0) {
            //Got the completed Transaction
            console.log("Transaction " + pendingInfo.tx + " confirmed in round " + pendingInfo.round);
            break;
        }
        await algodclient.statusAfterBlock(lastround + 1);
    }
};
async function writeUnsignedTransactionToFile() {

    try{
        const receiver = "GD64YIY3TWGDMCNPP553DZPPR6LDUSFQOIJVFDPPXWEG3FVOJCCDBBHU5A";

        // setup accounts and make node connection
        let algodClient = await setupClient();

        // recover account
        let myAccount = await recoverAccount();
        console.log("My address: %s", myAccount.addr)

        // get network suggested parameters
        let params = await algodClient.getTransactionParams();
        let txn = {
            "from": myAccount.addr,
            "to": receiver,
            "fee": params.minFee,
            "flatFee": true,
            "amount": 1000000,
            "firstRound": params.lastRound,
            "lastRound": params.lastRound + 1000,
            "genesisID": params.genesisID,
            "genesisHash": params.genesishashb64
        };
        // Save transaction to file
        fs.writeFileSync('./unsigned.txn', algosdk.encodeObj( txn ));   
    }catch( e ){
        console.log( e );
    }
}; 
async function readUnsignedTransactionFromFile() {

    try{
        // setup connection to node
        let algodClient = await setupClient();

        // recover account
        let myAccount = await recoverAccount(); 
        console.log("My address: %s", myAccount.addr)

        // read transaction from file and sign it
        let txn = algosdk.decodeObj(fs.readFileSync('./unsigned.txn'));  
        let signedTxn = algosdk.signTransaction(txn, myAccount.sk);
        let txId = signedTxn.txID;
        console.log("Signed transaction with txID: %s", txId);

        // send signed transaction to node
        await algodClient.sendRawTransaction(signedTxn.blob);

        // Wait for transaction to be confirmed
        await waitForConfirmation(algodClient, txId);
    } catch ( e ){
        console.log( e );
    }   
}; 
async function writeSignedTransactionToFile() {

    try{
        const receiver = "GD64YIY3TWGDMCNPP553DZPPR6LDUSFQOIJVFDPPXWEG3FVOJCCDBBHU5A";

        // setup connection to node
        let algodClient = await setupClient();
        let myAccount = await recoverAccount();
        console.log("My address: %s", myAccount.addr)

        // get network suggested parameters
        let params = await algodClient.getTransactionParams();

        // setup a transaction
        let txn = {
            "from": myAccount.addr,
            "to": receiver,
            "fee": params.minFee,
            "flatFee": true,
            "amount": 1000000,
            "firstRound": params.lastRound,
            "lastRound": params.lastRound + 1000,
            "genesisID": params.genesisID,
            "genesisHash": params.genesishashb64
        };

        // sign transaction and write to file
        let signedTxn = algosdk.signTransaction(txn, myAccount.sk);
        fs.writeFileSync('./signed.stxn', algosdk.encodeObj( signedTxn ));  
    } catch( e ) {
        console.log(e);
    }
}; 
async function readSignedTransactionFromFile() {

    try{
        // setup connection to node
        let algodClient = await setupClient();

        // read signed transaction from file
        let stx = algosdk.decodeObj(fs.readFileSync("./signed.stxn"));

        // send signed transaction to node
        let tx = await algodClient.sendRawTransaction(stx.blob);
        console.log("Signed transaction with txID: %s", tx.txId);

        // Wait for transaction to be confirmed
        await waitForConfirmation(algodClient, tx.txId);
    } catch( e ) {
        console.log(e);
    }   
}; 

async function testUnsigned(){
    await writeUnsignedTransactionToFile();
//    await readUnsignedTransactionFromFile();
}
async function testSigned(){
    await writeSignedTransactionToFile();
    await readSignedTransactionFromFile();
}
testUnsigned();
//testSigned();

File: examples/transactions/v1/python/offline_transaction_authorization.py

#/usr/bin/python3
import json
import time
import base64
import os
from algosdk import algod
from algosdk import mnemonic
from algosdk import transaction

def connect_to_network():
    algod_address = <algod-address>
    algod_token = <algod-token>
    algod_client = algod.AlgodClient(algod_token, algod_address)
    return algod_client

def wait_for_confirmation( algod_client, txid ):
    while True:
        txinfo = algod_client.pending_transaction_info(txid)
        if txinfo.get('round') and txinfo.get('round') > 0:
            print("Transaction {} confirmed in round {}.".format(txid, txinfo.get('round')))
            break
        else:
            print("Waiting for confirmation...")
            algod_client.status_after_block(algod_client.status().get('lastRound') +1)

def write_unsigned():
    # setup none connection
    algod_client = connect_to_network()

    # recover account
    passphrase = <25-word-passphrase>
    private_key = mnemonic.to_private_key(passphrase)
    my_address = mnemonic.to_public_key(passphrase)
    print("My address: {}".format(my_address))

    # get suggested parameters
    params = algod_client.suggested_params()

    # create transaction
    receiver = <transaction-receiver>
    data = {
        "sender": my_address,
        "receiver": receiver,
        "fee": params.get('minFee'),
        "flat_fee": True,
        "amt": <amount>,
        "first": params.get('lastRound'),
        "last": params.get('lastRound') + 1000,
        "gen": params.get('genesisID'),
        "gh": params.get('genesishashb64')
    }
    txn = transaction.PaymentTxn(**data)

    # write to file
    dir_path = os.path.dirname(os.path.realpath(__file__))
    transaction.write_to_file([txn], dir_path + "/unsigned.txn")


def read_unsigned():

    # setup node connection
    algod_client = connect_to_network()

    # recover account
    passphrase = "25-word-passphrase"
    private_key = mnemonic.to_private_key(passphrase)
    my_address = mnemonic.to_public_key(passphrase)
    print("My address: {}".format(my_address))

    # read from file
    txns = transaction.retrieve_from_file("./unsigned.txn")

    # sign and submit transaction
    txn = txns[0]
    signed_txn = txn.sign(private_key)
    txid = signed_txn.transaction.get_txid()
    print("Signed transaction with txID: {}".format(txid))
    algod_client.send_transaction(signed_txn)

    # wait for confirmation
    wait_for_confirmation( algod_client, txid)

def write_signed():

    # setup connection to node
    algod_client = connect_to_network()

    # recovere account
    passphrase = <25-word-passphrase>
    private_key = mnemonic.to_private_key(passphrase)
    my_address = mnemonic.to_public_key(passphrase)
    print("My address: {}".format(my_address))

    # get node suggested parameters
    params = algod_client.suggested_params()

    # create transaction
    receiver = <transaction-receiver>
    data = {
        "sender": my_address,
        "receiver": receiver,
        "fee": params.get('minFee'),
        "flat_fee": True,
        "amt": <amount>,
        "first": params.get('lastRound'),
        "last": params.get('lastRound') + 1000,
        "gen": params.get('genesisID'),
        "gh": params.get('genesishashb64')
    }
    txn = transaction.PaymentTxn(**data)

    # sign transaction
    signed_txn = txn.sign(private_key)

    # write to file
    dir_path = os.path.dirname(os.path.realpath(__file__))
    transaction.write_to_file([signed_txn], dir_path + "/signed.txn")


def read_signed():

    # set up connection to node
    algod_client = connect_to_network()

    # read signed transaction from file
    txns = transaction.retrieve_from_file("./signed.txn")
    signed_txn = txns[0]
    txid = signed_txn.transaction.get_txid()
    print("Signed transaction with txID: {}".format(txid))

    # send transaction to network
    algod_client.send_transaction(signed_txn)

    # wait for confirmation
    wait_for_confirmation( algod_client, txid)

# Test Runs     
#write_unsigned()
#read_unsigned()
write_signed()
read_signed()    